File: airbyte-connector-builder-server/src/test/java/io/airbyte/connector_builder/controllers/ConnectorBuilderControllerTest.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.connector_builder.controllers;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import io.airbyte.connector_builder.api.model.generated.ResolveManifest;
import io.airbyte.connector_builder.api.model.generated.ResolveManifestRequestBody;
import io.airbyte.connector_builder.api.model.generated.StreamRead;
import io.airbyte.connector_builder.api.model.generated.StreamReadRequestBody;
import io.airbyte.connector_builder.exceptions.AirbyteCdkInvalidInputException;
import io.airbyte.connector_builder.handlers.AssistProxyHandler;
import io.airbyte.connector_builder.handlers.ConnectorContributionHandler;
import io.airbyte.connector_builder.handlers.HealthHandler;
import io.airbyte.connector_builder.handlers.ResolveManifestHandler;
import io.airbyte.connector_builder.handlers.StreamHandler;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class ConnectorBuilderControllerTest {

  private ConnectorBuilderController controller;
  private HealthHandler healthHandler;
  private StreamHandler streamHandler;
  private ResolveManifestHandler resolveManifestHandler;
  private StreamReadRequestBody streamReadRequestBody;
  private StreamRead streamReadResponse;
  private ResolveManifestRequestBody resolveManifestRequestBody;
  private ResolveManifest resolveManifest;
  private ConnectorContributionHandler connectorContributionHandler;
  private AssistProxyHandler assistProxyHandler;

  @BeforeEach
  void setup() {
    this.healthHandler = mock(HealthHandler.class);
    this.resolveManifestHandler = mock(ResolveManifestHandler.class);
    this.streamHandler = mock(StreamHandler.class);

    this.streamReadRequestBody = mock(StreamReadRequestBody.class);
    this.streamReadResponse = mock(StreamRead.class);
    this.resolveManifestRequestBody = mock(ResolveManifestRequestBody.class);
    this.resolveManifest = mock(ResolveManifest.class);
    this.connectorContributionHandler = mock(ConnectorContributionHandler.class);
    this.assistProxyHandler = mock(AssistProxyHandler.class);

    this.controller =
        new ConnectorBuilderController(this.healthHandler, this.resolveManifestHandler, this.streamHandler, this.connectorContributionHandler,
            this.assistProxyHandler);
  }

  @Test
  void whenReadStreamThenReturnHandlerResponse() {
    when(streamHandler.readStream(streamReadRequestBody)).thenReturn(streamReadResponse);
    final StreamRead response = this.controller.readStream(streamReadRequestBody);
    assertEquals(streamReadResponse, response);
  }

  @Test
  void givenExceptionWhenReadStreamThenThrowSameException() {
    when(streamHandler.readStream(any())).thenThrow(AirbyteCdkInvalidInputException.class);
    assertThrows(AirbyteCdkInvalidInputException.class, () -> this.controller.readStream(streamReadRequestBody));
  }

  @Test
  void whenResolveManifestThenReturnHandlerResponse() {
    when(resolveManifestHandler.resolveManifest(resolveManifestRequestBody)).thenReturn(resolveManifest);
    final ResolveManifest response = this.controller.resolveManifest(resolveManifestRequestBody);
    assertEquals(resolveManifest, response);
  }

  @Test
  void givenExceptionWhenResolveManifestThenThrowSameException() {
    when(resolveManifestHandler.resolveManifest(resolveManifestRequestBody)).thenThrow(AirbyteCdkInvalidInputException.class);
    assertThrows(AirbyteCdkInvalidInputException.class, () -> this.controller.resolveManifest(resolveManifestRequestBody));
  }

}

File: airbyte-connector-builder-server/src/test/kotlin/io/airbyte/connectorbuilder/controllers/ConnectorBuilderControllerTest.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.connectorbuilder.controllers

import io.airbyte.commons.server.handlers.AssistProxyHandler
import io.airbyte.commons.server.handlers.ConnectorContributionHandler
import io.airbyte.connectorbuilder.api.model.generated.ResolveManifest
import io.airbyte.connectorbuilder.api.model.generated.ResolveManifestRequestBody
import io.airbyte.connectorbuilder.api.model.generated.StreamRead
import io.airbyte.connectorbuilder.api.model.generated.StreamReadRequestBody
import io.airbyte.connectorbuilder.exceptions.AirbyteCdkInvalidInputException
import io.airbyte.connectorbuilder.handlers.FullResolveManifestHandler
import io.airbyte.connectorbuilder.handlers.HealthHandler
import io.airbyte.connectorbuilder.handlers.ResolveManifestHandler
import io.airbyte.connectorbuilder.handlers.StreamHandler
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

internal class ConnectorBuilderControllerTest {
  private lateinit var controller: ConnectorBuilderController
  private var healthHandler: HealthHandler = mockk()
  private var streamHandler: StreamHandler = mockk()
  private var resolveManifestHandler: ResolveManifestHandler = mockk()
  private var fullResolveManifestHandler: FullResolveManifestHandler = mockk()
  private var streamReadRequestBody: StreamReadRequestBody = mockk()
  private var streamReadResponse: StreamRead = mockk()
  private var resolveManifestRequestBody: ResolveManifestRequestBody = mockk()
  private var resolveManifest: ResolveManifest = mockk()
  private var connectorContributionHandler: ConnectorContributionHandler = mockk()
  private var assistProxyHandler: AssistProxyHandler = mockk()

  @BeforeEach
  fun setup() {
    clearAllMocks()

    controller =
      ConnectorBuilderController(
        healthHandler = healthHandler,
        resolveManifestHandler = resolveManifestHandler,
        fullResolveManifestHandler = fullResolveManifestHandler,
        streamHandler = streamHandler,
        connectorContributionHandler = connectorContributionHandler,
        assistProxyHandler = assistProxyHandler,
      )
  }

  @Test
  fun whenReadStreamThenReturnHandlerResponse() {
    every { streamHandler.readStream(streamReadRequestBody) } answers { streamReadResponse }
    val response = controller.readStream(streamReadRequestBody)
    assertEquals(streamReadResponse, response)
  }

  @Test
  fun givenExceptionWhenReadStreamThenThrowSameException() {
    every { streamHandler.readStream(any()) } throws AirbyteCdkInvalidInputException("message")
    assertThrows<AirbyteCdkInvalidInputException> { controller.readStream(streamReadRequestBody) }
  }

  @Test
  fun whenResolveManifestThenReturnHandlerResponse() {
    every { resolveManifestHandler.resolveManifest(resolveManifestRequestBody) } answers { resolveManifest }
    val response = controller.resolveManifest(resolveManifestRequestBody)
    assertEquals(resolveManifest, response)
  }

  @Test
  fun givenExceptionWhenResolveManifestThenThrowSameException() {
    every { resolveManifestHandler.resolveManifest(any()) } throws AirbyteCdkInvalidInputException("message")
    assertThrows<AirbyteCdkInvalidInputException> {
      controller.resolveManifest(resolveManifestRequestBody)
    }
  }
}
